src/defender_metrics/mod.rs

```diff
@@ -0,0 +1,404 @@
+use crate::shadows::Error;
+use data_types::Metric;
+use embassy_sync::blocking_mutex::raw::RawMutex;
+use embedded_mqtt::{DeferredPayload, Publish, Subscribe, SubscribeTopic, ToPayload};
+use errors::{ErrorResponse, MetricError};
+use futures::StreamExt;
+use serde::Serialize;
+use topics::Topic;
+
+// pub mod aws_types;
+pub mod aws_types;
+pub mod data_types;
+pub mod errors;
+pub mod topics;
+
+pub struct MetricHandler<'a, 'm, M: RawMutex> {
+    mqtt: &'m embedded_mqtt::MqttClient<'a, M>,
+}
+
+impl<'a, 'm, M: RawMutex> MetricHandler<'a, 'm, M> {
+    pub fn new(mqtt: &'m embedded_mqtt::MqttClient<'a, M>) -> Self {
+        Self { mqtt }
+    }
+
+    pub async fn publish_metric<'c, C: Serialize>(
+        &self,
+        metric: Metric<'c, C>,
+        max_payload_size: usize,
+    ) -> Result<(), MetricError> {
+        let payload = DeferredPayload::new(
+            |buf: &mut [u8]| {
+                #[cfg(feature = "metric_cbor")]
+                {
+                    let mut serializer = minicbor_serde::Serializer::new(
+                        minicbor::encode::write::Cursor::new(&mut *buf),
+                    );
+
+                    match metric.serialize(&mut serializer) {
+                        Ok(_) => {}
+                        Err(_) => {
+                            error!("An error happened when serializing metric with cbor");
+                        }
+                    };
+
+                    Ok(serializer.into_encoder().writer().position())
+                }
+
+                #[cfg(not(feature = "metric_cbor"))]
+                {
+                    serde_json_core::to_slice(&metric, buf)
+                        .map_err(|_| embedded_mqtt::EncodingError::BufferSize)
+                }
+            },
+            max_payload_size,
+        );
+
+        //Wait for mqtt to connect
+        self.mqtt.wait_connected().await;
```

Comment:
Could you move this above the DeferredPayload? That way you don't need to hold onto the closure across an await point that could potentially be a while, depending on network situation. Compiler might not be able to optimize.